### DISCUSSION TITLE
Does APSIM see all CPUs available on a HPC node or just those requested?

### DISCUSSION THREAD

**Discussion by BrianCollinss** [APSIMInitiative/ApsimX#8813] [Q&A]
We have found that some APSIM jobs on the HPC take significantly longer than others to complete. Our HPC admin has informed me that jobs that ran slowly were all on the same compute node and jobs that ran quickly seem to be the only jobs running on a compute node:

> ... Historically, APSIM uses all the resources available on a compute node. As an example, you ask for 41 cores for a job but Models sees 128 cores on a compute node so it uses all 128 cores.

The admin recommends we use the Models "—cpu-count" option to address the issue. I don’t force APSIM’s Models to use a specific number of CPUs; however, I do set two other variables when I run APSIM (along with #PBS -l ncpus=XX):

export NUMBER_OF_PROCESSORS=XX
export SINGULARITYENV_NUMBER_OF_PROCESSORS=XX-1

I wonder if you concur with our admin's comment on APSIM's tendency to use ALL existing CPUs, and whether you also recommend using the “—cpu-count” option?

**Comment by byzheng** [comment]
Hi Brain, I would suggest 

* constraining APSIM uses a single CPU on cluster
* reducing I/O  

## Single CPU

Although I don't test the difference of performance. I believe the overhead is to load models and runtime libraries into memory. The overhead can be ignored for longer runtime of single apsimx file.

## I/O


Depending on how you store files (i.e. apsimx, met, db), the major bottleneck is I/O of db file by sqlite3 library. APSIM NG cached 100 rows in a table in default or end of simulation to start a transaction of db. I/O will be trouble if apsimx file contains lots of single season simulations with yearly report. 

If all files are stored in a single physical storage (e.g. NFS), some processes will be much slower than others if there are too much file connections for frequent writing.  #8759 is an example for runtime ranged from 80 to 250 s to write all files into shared high speed hard drive. 

One workaround would be to save apsimx into local temp folder in compute nodes (i.e. /tmp), run simulations with I/O in local drive, and finally copy output back to shared storage.

Also disable all information message to reduce file size of db (#8480)

**Reply by BrianCollinss** [reply]
Thanks Bangyou. I do most of the things you mentioned, including running APSIM in the temp folder of the running node. The issue is that my jobs are almost identical but some took 3x longer to complete. When I re-ran those slow jobs, they were run like the others, implying something went wrong on some running machines.

**Comment by BrianCollinss** [comment]
Hey @peter-devoil and @hol353. Do you have any comment or have you had similar experience?

**Comment by peter-devoil** [comment]
No - I haven't seen that, though BZ's comment about IO (can give you grief if you're doing lots of output) may be relevant if the offending node is memory constrained and starts swapping.

Remember that you may be sharing the silicon with other jobs that thrash cache lines etc - so as long as you are running a (factorial simulation?) with many simulations run in parallell you are quite justified in grabbing all the cores you can. The slow bit in apsim is resolving links before the start of simulation, which is partially sidestepped in a factorial simln.

I cant see NG using the PBS environment variables, only --cpu-count args

**Reply by BrianCollinss** [reply]
Thanks Peter. My assumption was that by setting the env variables for the container, APSIM will only see the number of CPUs set for the container to use. In other words, can/should APSIM be able to see all the CPUs available outside the container?

**Comment by peter-devoil** [comment]
No. Apsim uses an  [dotnet environment variable](https://learn.microsoft.com/en-us/dotnet/api/system.environment.processorcount)  (which in turn likely queries /proc/cpuinfo) that is ignorant of containers and PBS conventions. See APSIM.Shared/JobRunning/JobRunner.cs:63

**Comment by byzheng** [comment]
As Peter mentioned, make sure you are not memory limitation. Recently, I accidentally utilise 30 MB memory to run APSIM NG in our cluster. The runtime ranges from 1 min to 30 min.

Another possible reason is to simultaneously load APSIMX executable file and libraries from all processes in HPC especially if you store them in a shared drive though NFS.  I generally pause my scripts with a random time  (0-30 s) before calling `Models` or other I/O process.

**Comment by BrianCollinss** [comment]
Thanks Peter and Bangyou. All the reasons you mentioned make sense to me. The thing is I wonder why it only happens when APSIM is being run on a node by itself. That makes me be more inclined towards APSIM taking too many CPUs while its other instance is trying to do the same. I assume the safe approach here is to always use --cpu flag with Models.

**Reply by byzheng** [reply]
I believe you could write a manager script to output number of processors

```
Environment.GetEnvironmentVariable("NUMBER_OF_PROCESSORS")
```
